singularity-eos/eos/vector_utils.hpp

```diff
@@ -0,0 +1,74 @@
+//------------------------------------------------------------------------------
+// © 2021. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_VECTOR_UTILS_
+#define _SINGULARITY_EOS_EOS_VECTOR_UTILS_
+
+#include <ports-of-call/portability.hpp>
+
+namespace singularity {
+namespace vector_utils {
+
+/*
+Class where the [] operator returns the null pointer for any index
+*/
+class NullIndexer {
+ public:
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real *operator[](int i) {
+    return nullptr;
+  }
+};
+
+/*
+Wrappers for scalar lookups to allow them to take vectors on the CPU but then
+call GPU kernels to do the actual lookups
+
+RealIndexer must have an operator[](int) that returns a Real. e.g., Real*
+ConstRealIndexer is as RealIndexer, but assumed const type.
+LambdaIndexer must have an operator[](int) that returns a Real*. e.g., Real**
+*/
+template<typename EOS, typename LookupFunction, typename ConstRealIndexer,
+         typename RealIndexer, typename LambdaIndexer>
+inline
+void VectorizeScalarLookup(const EOS &eos, const LookupFunction &&f,
+                           ConstRealIndexer &&xs, ConstRealIndexer &&ys,
+                           RealIndexer &&result, const int num,
+                           LambdaIndexer &&lambdas) {
+  portableFor(
```

Comment:
@jhp-lanl perhaps adding a `const char*` for a function name may be helpful here. Another option I've been toying with is to grab additional info from the template parameters, but getting the name of the specific function passed in so far has been challenging. I propose:
- passing in the string (lots of work added, but provides control and lots of information)
- using template param(s) (little work, but less specific, not sure if typeid overhead is perceived on every function call)

Reply:
I like the string solution better. Shouldn't be too bad.

Reply:
I wrote an example of how a macro could be used that may be less objectionable to @jhp-lanl but still provide the specificity we need:
```c++
template<typename T, typename U>
void print_name(const T& a, U&& b, const char* s) {
    printf("%s\n", s);
}

#define PRINT_NAME(a, b) \
print_name(a, b, #b)

int main() {
    PRINT_NAME(gruneisen{}, &gruneisen::funcone); // string is "&gruneisen::funcone"
    return 0;
}
```

Reply:
Nice.

Reply:
Sorry I'm a little confused. What did you want me to do with the function name?

Reply:
Resolving because I changed everything lol